// myModules/resultMods.js
var exports = module.exports = {};

// Get single result
exports.getResult = function(req, query, callback) {
    var db = req.db;
    var collection = db.get('results');
    collection.findOne(query,{},function(e,doc){
        if(e || doc === null){
            callback({});
        }
        else {
            callback(doc);
        };
    });
}

// Get list of results
exports.getresults = function(query, req, callback) {
    var db = req.db;
    var results = [];
    
    var collection = db.get('tests');
    collection.find(query,{},function(e,docs){
        results = docs;
        if(e || docs === null){
            callback([]);
        }
        else {
            callback(docs);
        }
    });
}
// Get list of results with related test appended to each result
exports.getResultsWithTest = function(req, query, withNotes, callback) {
    options = {
        _id: 1,
        scenario_id: 1,
        test_id: 1,
        result: 1,
        message: 1,
        duration: 1,
    }
    if(withNotes){
       options['notes'] = 1 
    }
    var db = req.db;
    var collection = db.get('results');
    collection.col.aggregate([
        {$project: options},
        {$match: query},
        {$lookup: {from: "tests",localField: "test_id",foreignField: "_id", as: "test"}},
        {$unwind: "$test"},
        {$sort: {_id: -1}}], function(e,docs){
        callback(docs);
    });
}

// Get result with appended scenario
exports.getTestResult = function(req, callback) {
    var db = req.db;
    var resultId = req.params.id;
    var ObjectId = require('mongodb').ObjectID;
    var resultsCollection = db.get('results');
    resultsCollection.col.aggregate([
        {$match: {_id: new ObjectId(resultId)}},
        {$lookup: {from: "scenarios",localField: "scenario_id",foreignField: "_id", as: "scenario"}},
        {$unwind: "$scenario"}], function(e,docs){
        callback(docs[0]);
    });
}
